### PR TITLE
Scrape ceased members with links to member pages

### PR DESCRIPTION
**Pull request by ondenman**
This is a step towards fixing: https://github.com/everypolitician/everypolitician-data/issues/25491

This scrapes three members who are no longer part of the legislature and are now listed on a separate page which has a layout that is different from the party group listings pages. There are fourteen members listed on this page but only three of them have links to individual member pages. The layout of the page is quite complex but these three members are easier to capture than the others.

Since, the three members are going missing in the latest import PRs, it makes sense to deal with them in their own PR. (issue: https://github.com/everypolitician-scrapers/macedonia-sobranie/issues/8)

### CODE REVIEW

File: scraper.rb [hunk 1]
```diff
@@ -21,11 +21,19 @@ def scrape(h)
 ScraperWiki.sqliteexecute('DELETE FROM data') rescue nil
 
 start = 'http://sobranie.mk/current-structure-2014-2018.nspx'
-data = scrape(start => GroupsPage).groups.flat_map do |group|
+ceased_members_url = 'http://sobranie.mk/mps-whose-mandate-has-not-been-completed-2014-2018.nspx'
```
Comment (tmtmtmtm) [inline scraper.rb]: Hard-coding this here and also in the `:groups` field is a warning sign. If this is something that appears on that page, then that should be a field on the page, which you can ask for here.

File: lib/groups_page.rb [hunk 2]
```diff
@@ -5,8 +5,9 @@ class GroupsPage < SobraniePage
   decorator Scraped::Response::Decorator::AbsoluteUrls
 
   field :groups do
-    noko.css('div.toc li a').map do |a|
-      fragment a => LinkLine
-    end
+    # The rejected page is not a list of members grouped by party
+    noko.css('div.toc li a')
+        .reject { |a| a.attr('href') == 'http://sobranie.mk/mps-whose-mandate-has-not-been-completed-2014-2018.nspx' }
```
Comment (tmtmtmtm) [inline lib/groups_page.rb]: Comparing this against the full URL like this is a mild warning sign: you're increasing the number of things that could change causing this to break. If there's no better way to distinguish the link than by examining the URL, I'd at least restrict it to just the 'mps-whose-mandate-has-not-been-completed' part, so that, for example, switching the site to https: wouldn't cause this to fail unnecessarily.

File: scraper.rb [hunk 3]
```diff
@@ -21,11 +21,19 @@ def scrape(h)
...
   end
 end
-# puts data.map { |r| r.reject { |k, v| v.to_s.empty? }.sort_by { |k,v| k }.to_h }
 
+ceased_members = scrape(ceased_members_url => CeasedMembersPage).members.map do |mem|
+  mem.to_h.merge(scrape(mem.source => MemberPage).to_h.merge(party: mem.party, term: term))
+end
+
+# puts data.map { |r| r.reject { |k, v| v.to_s.empty? }.sort_by { |k,v| k }.to_h }
+data = (current_members << ceased_members).flatten
```
Comment (tmtmtmtm) [inline scraper.rb]: why mutate `current_members` here?

File: lib/ceased_members_page.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+require_relative 'sobranie_page'
+require_relative 'ceased_member_row'
+
+class CeasedMembersPage < SobraniePage
+  field :members do
+    noko.xpath('//div[@class="row"]/p[a]').map do |p|
```
Comment (tmtmtmtm) [inline lib/ceased_members_page.rb:7]: Why are you restricting this to only the final 3 out of the 14 members on this page?

File: lib/ceased_member_row.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  field :name do
+    noko.css('a').text
+  end
+
+  field :party do
+    # The layout is inconsistent. Scraping this way
+    # we always get the party name for each row
+    noko.xpath('text()')[0..1].text.split('.').last.tidy
```
Comment (tmtmtmtm) [inline lib/ceased_member_row.rb]: This seems over complex, and I'm not sure what the variation is in layout that you're seeing (at least between the members that you're looking at). 

The tests pass with `noko.xpath('text()[2]').text.tidy` (though that's merely pointing out a simplification of what you're doing, not suggesting that that's a good way to do this)

File: lib/ceased_member_row.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  field :end_date do
+    Date.parse(noko.text.match(/(?:\d\d?\.\d\d?.\d\d\d\d)/)[0]).to_s
```
Comment (tmtmtmtm) [inline lib/ceased_member_row.rb]: 1. I can't work out why you're starting this with a non-capturing `?:`
2. You're quoting the '.' after the first group of digits, but not the second one: this should be consistent
2. A big chain of `\d`s like this, especially where some of them are actually `\d?` is quite hard to read. Being explicit with counts might be more legible: `\d{1,2}\.\d{1,2}\.\d{4}`
3. `text.match(/…/)[0]` is a little easier to follow (and is certainly more consistent with how we do it elsewhere), as `text[/…/, 1]`

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

File: lib/ceased_member_row.rb (hunk 5 above)
Comment (ondenman) [reply lib/ceased_member_row.rb]: Silvana Boneva's row is slightly different from the others. In the final three rows, the party ID is either in the first or second text nodes. I've added a test for Silvana Boneva.

Comment (tmtmtmtm) [reply lib/ceased_member_row.rb]: 👍  — always important to have an explicit test case for the edge-case.

File: lib/ceased_members_page.rb (hunk 4 above)
Comment (ondenman) [reply lib/ceased_members_page.rb:7]: I've opened an [issue](https://github.com/everypolitician-scrapers/macedonia-sobranie/issues/8) and have changed the PR description.

File: lib/ceased_member_row.rb (hunk 5 above)
Comment (tmtmtmtm) [reply lib/ceased_member_row.rb]: This is still odd on multiple fronts.  Calling `[0..1]` is always a bit of warning sign — that's usually simpler as `.take(2)`. But here you're calling `text` on the result, which is even more odd — especially as you're then calling `.split.last` to get rid of the first one again. 

So you could simplify all of this as:

```ruby
    noko.xpath('text()').take(2).last.text.tidy
```

However, that's probably not as good as being more expressive about what you actually want, which is the first text element after the linked name, which would be:

```ruby
    noko.xpath('a/following-sibling::text()').first.text.tidy
```

File: lib/groups_page.rb [hunk 7]
```diff
@@ -4,9 +4,21 @@
 class GroupsPage < SobraniePage
   decorator Scraped::Response::Decorator::AbsoluteUrls
 
+  CEASED_MEMBERS_URL = 'mps-whose-mandate-has-not-been-completed-2014-2018'
```
Comment (tmtmtmtm) [inline lib/groups_page.rb]: Hard-coding the dates here means that this page will only work for this single term, even though this should be a class that works for all historic terms too…

File: lib/groups_page.rb [hunk 8]
```diff
@@ -4,9 +4,21 @@
...
-      fragment a => LinkLine
-    end
+    # The rejected page is not a list of members grouped by party
+    groups_links.reject { |a| a.attr('href').include? CEASED_MEMBERS_URL }
+                .map { |a| fragment a => LinkLine }
+  end
+
+  field :ceased_members_url do
+    groups_links.find { |a| a.attr('href').include? CEASED_MEMBERS_URL }.attr('href')
+  end
```
Comment (tmtmtmtm) [inline lib/groups_page.rb]: There's a little too much repetition here. I'd suggest instead that you `partition` on that, along the lines of:

```ruby
  field :groups do
    groups_links.last.map { |a| fragment a => LinkLine }
  end

  field :ceased_members_url do
    groups_links.first.first.attr('href')
  end

  private

  # The layout of this page is to list all the groups, but at the end,
  # in an otherwise undistinguished entry, there's a list to a separate
  # page of all the ceased members. 
  def groups_links
    noko.css('div.toc li a').partition { |a| a.attr('href').include? CEASED_MEMBERS_URL }
  end
```

File: scraper.rb [hunk 9]
```diff
@@ -21,11 +21,20 @@ def scrape(h)
...
   end
 end
-# puts data.map { |r| r.reject { |k, v| v.to_s.empty? }.sort_by { |k,v| k }.to_h }
 
+ceased_members = scrape(groups_page.ceased_members_url => CeasedMembersPage).members.map do |mem|
+  mem.to_h.merge(scrape(mem.source => MemberPage).to_h.merge(party: mem.party, term: term))
+end
+
+# puts data.map { |r| r.reject { |k, v| v.to_s.empty? }.sort_by { |k,v| k }.to_h }
+data = current_members + ceased_members
```
Comment (tmtmtmtm) [inline scraper.rb]: the commented out line is referring to a variable that doesn't exist yet. These should really be the other way around.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

**Review by tmtmtmtm — APPROVED** [review]
This could probably have done with some more comments in the code itself to explain that it's semi-deliberately only working with the last three people, but 👍
